### PR TITLE
fix: resolved border collapse issue #1626 when border variants are mixed in button group

### LINKED ISSUE #1626: [BUG] - Button Group issues
### NextUI Version

2.0.15

### Describe the bug

`variant="bordered"` does not show correct borders, the selected tab should have all borders

Bildschirmfoto 2023-09-14 um 21 40 21

Also setting `radius` on the `ButtonGroup` does not affect the buttons

### Your Example Website or App

_No response_

### Steps to Reproduce the Bug or Issue

Create a button group with buttons, try to set `radius` or use bordered variant

### Expected behavior

see above

### Screenshots or Videos

_No response_

### Operating System Version

mac0s

### Browser

Firefox

### PR DESCRIPTION
**Pull request by chongruei**
Closes #1626

## 📝 Description

The ```radius``` property does not work on the 'buttonGroup' component. The ```variant="bordered"``` option does not display the correct borders.

## ⛳️ Current behavior (updates)

The ```radius``` is always the same as ```md```.
The ```border-left``` is covered by  margin-left ```ml-[calc(theme(borderWidth.medium)*-1)]```.

```tsx
  <ButtonGroup color="success" radius="lg" variant="flat">
    <Button>One</Button>
    <Button>Two</Button>
    <Button variant="bordered">Three</Button>
    <Button>Four</Button>
    <Button>Five</Button>
    <Button>Six</Button>
  </ButtonGroup>
```

## 🚀 New behavior

The ```radius``` now works, and the border is displayed correctly in the ```buttonGroup```.

## 💣 Is this a breaking change (Yes/No):

No

## 📝 Additional Information

Please kindly let me know if any suggestions.

### CODE REVIEW

File: .changeset/chilly-trainers-type.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+"@nextui-org/button": patch
+"@nextui-org/theme": patch
+---
+
+Fix #1626 The radius now works, and the border is displayed correctly in the buttonGroup.
```
Comment (jguddas) [inline .changeset/chilly-trainers-type.md:6]: Can you describe the change more precisely, "now works" does not tell me anything?

**Comment by jguddas** [comment]
Looks pretty broken:
image

https://nextui-storybook-v2-git-fork-chongruei-fix162-f030f9-nextui-org.vercel.app/?path=/story/components-buttongroup--default

**Comment by Digoler** [comment]
I am a new frontend developer who wants to improve programming skills through open source. I would like to know how you solved this problem. Could you please tell me the reasons and solutions for this problem?

**Comment by jguddas** [comment]
> I am a new frontend developer who wants to improve programming skills through open source. I would like to know how you solved this problem. Could you please tell me the reasons and solutions for this problem?

It wasn't solved.

File: .changeset/chilly-trainers-type.md (hunk 1 above)
Comment (chongruei) [reply .changeset/chilly-trainers-type.md]: sorry, @jguddas  can you check this again? thank you.

**Comment by chongruei** [comment]
sorry, @jguddas I've missed radius as the default prop from theme.ts.

**Comment by jguddas** [comment]
Can you add something like this as a story?

```
  <ButtonGroup color="success" radius="lg" variant="flat">
    <Button>One</Button>
    <Button>Two</Button>
    <Button variant="bordered">Three</Button>
    <Button>Four</Button>
    <Button>Five</Button>
    <Button>Six</Button>
  </ButtonGroup>
```

**Comment by Digoler** [comment]
> Hi @jguddas
> 
> Sorry, I've missed radius as the default prop from theme.ts. Could you check this again?

Can you tell me why you added an isIsolate attribute instead of directly taking the following approach?

image

**Comment by Digoler** [comment]
> > Hi @jguddas
> > Sorry, I've missed radius as the default prop from theme.ts. Could you check this again?
> 
> Can you tell me why you added an isIsolate attribute instead of directly taking the following approach?
> 
> image

This can also solve this problem.

File: packages/components/button/stories/button-group.stories.tsx [hunk 2]
```diff
@@ -63,10 +63,26 @@ const Template = (args: ButtonGroupProps) => (
   </ButtonGroup>
 );
 
+const VariantButtonTemplate = (args: ButtonGroupProps) => (
+  <ButtonGroup {...args} variant="solid">
```
Comment (jguddas) [inline packages/components/button/stories/button-group.stories.tsx:67]:
```suggestion
  <ButtonGroup {...args}>
```

File: packages/components/button/stories/button-group.stories.tsx [hunk 3]
```diff
@@ -63,10 +63,26 @@ const Template = (args: ButtonGroupProps) => (
...
   args: {
     ...defaultProps,
   },
 };
+
+export const VariantButton = {
+  render: VariantButtonTemplate,
+
+  args: {
+    ...defaultProps,
```
Comment (jguddas) [inline packages/components/button/stories/button-group.stories.tsx:86]:
```suggestion
    ...defaultProps,
    variant: "solid",
```

**Review by jguddas — CHANGES_REQUESTED** [review]

File: .changeset/chilly-trainers-type.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+"@nextui-org/button": patch
+"@nextui-org/theme": patch
+---
+
+Fix #1626 The 'radius' will be changed based on the ButtonGroup props. However, the 'border-left' is obscured by 'margin-left ml-[calc(theme(borderWidth.medium)*-1)]', and the border is not covered by its neighbor when the button is set to variant='bordered' in the ButtonGroup.
```
Comment (jguddas) [inline .changeset/chilly-trainers-type.md:6]:
```suggestion
Added support for setting `variant` on `Button` element inside `ButtonGroup` to `"bordered"`.
Resolves #1626

The 'radius' will be changed based on the ButtonGroup props. However, the 'border-left' is obscured by 'margin-left ml-[calc(theme(borderWidth.medium)*-1)]', and the border is not covered by its neighbor when the button is set to variant='bordered' in the ButtonGroup.
```

**Comment by chongruei** [comment]
> > > Hi @jguddas
> > > Sorry, I've missed radius as the default prop from theme.ts. Could you check this again?
> > 
> > 
> > Can you tell me why you added an isIsolate attribute instead of directly taking the following approach?
> > image
> 
> This can also solve this problem.

Hi, @Digoler, it would be like without isIsolate

Screenshot 2023-09-17 at 1 26 04 AM

After defining the isIsolate

Screenshot 2023-09-17 at 1 29 51 AM

File: packages/core/theme/src/components/button.ts [hunk 5]
```diff
@@ -316,31 +319,36 @@ const button = tv({
...
       isInGroup: true,
-      size: "lg",
+      radius: "lg",
       class: "rounded-none first:rounded-l-large last:rounded-r-large",
     },
     {
       isInGroup: true,
-      isRounded: true,
+      radius: "full",
       class: "rounded-none first:rounded-l-full last:rounded-r-full",
```
Comment (jguddas) [inline packages/core/theme/src/components/button.ts:345]: But this isn't really related, is it, could we put this into a separate PR?

File: packages/components/button/stories/button-group.stories.tsx [hunk 6]
```diff
@@ -63,10 +63,26 @@ const Template = (args: ButtonGroupProps) => (
   </ButtonGroup>
 );
 
+const VariantButtonTemplate = (args: ButtonGroupProps) => (
+  <ButtonGroup {...args} variant="solid">
+    <Button>One</Button>
+    <Button variant={args.variant}>Two</Button>
```
Comment (jguddas) [inline packages/components/button/stories/button-group.stories.tsx:69]:
```suggestion
    <Button variant="bordered">Two</Button>
```

File: packages/components/button/src/use-button.ts [hunk 7]
```diff
@@ -112,6 +113,7 @@ export function useButton(props: UseButtonProps) {
         fullWidth,
         isDisabled,
         isInGroup,
+        isIsolate,
```
Comment (jguddas) [inline packages/components/button/src/use-button.ts:116]: Should this not be `isIsolated` instead of `isIsolate`?
```suggestion
        isIsolated,
```

**Comment by Digoler** [comment]
> > > > Hi @jguddas
> > > > Sorry, I've missed radius as the default prop from theme.ts. Could you check this again?
> > > 
> > > 
> > > Can you tell me why you added an isIsolate attribute instead of directly taking the following approach?
> > > image
> > 
> > 
> > This can also solve this problem.
> 
> Hi, @Digoler, it would be like without isIsolate
> 
> Screenshot 2023-09-17 at 1 26 04 AM
> 
> After defining the isIsolate
> 
> Screenshot 2023-09-17 at 1 29 51 AM

Thank you, I have learned a lot.

File: packages/components/button/src/use-button.ts (hunk 7 above)
Comment (chongruei) [reply packages/components/button/src/use-button.ts:116]: Looks better!

File: packages/core/theme/src/components/button.ts (hunk 5 above)
Comment (chongruei) [reply packages/core/theme/src/components/button.ts:345]: Sure! I can recover the radius part for another PR.

**Comment by jguddas** [comment]
Another edge case for you

```tsx
  <ButtonGroup color="success" radius="lg" variant="flat">
    <Button>One</Button>
    <Button>Two</Button>
    <Button variant="bordered">Three</Button>
    <Button variant="bordered">Four</Button>
    <Button variant="bordered">Five</Button>
    <Button>Six</Button>
  </ButtonGroup>
```

**Comment by jguddas** [comment]
And another one where `ButtonGroup.variant` is `bordered`

```tsx
 <ButtonGroup color="success" radius="lg" variant="bordered">
    <Button>One</Button>
    <Button>Two</Button>
    <Button variant="bordered">Three</Button>
    <Button>Four</Button>
    <Button>Five</Button>
    <Button>Six</Button>
  </ButtonGroup>
```

**Comment by chongruei** [comment]
> And another one where `ButtonGroup.variant` is `bordered`
> 
> ```tsx
>  <ButtonGroup color="success" radius="lg" variant="bordered">
>     <Button>One</Button>
>     <Button>Two</Button>
>     <Button variant="bordered">Three</Button>
>     <Button>Four</Button>
>     <Button>Five</Button>
>     <Button>Six</Button>
>   </ButtonGroup>
> ```

Thank you! I will check for this.

**Comment by jguddas** [comment]
In case you are closing the PR, so it does not get accidentally merged:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

**Comment by chongruei** [comment]
Hi! @jguddas 

I finally removed the negative value for the margin because it always covered the neighbor.

Like this 
Screenshot 2023-09-18 at 1 12 34 AM

It appears that removing the negative value for the margin has led to a situation where the border size doubles when the colors are the same. What are your thoughts on this issue?
Screenshot 2023-09-18 at 1 20 43 AM

**Comment by jguddas** [comment]
> It appears that removing the negative value for the margin has led to a situation where the border size doubles when the colors are the same. What are your thoughts on this issue? Screenshot 2023-09-18 at 1 20 43 AM

Yeah, not a great look.

**My proposal:** keep the `isIsolated` but only have it be true when `groupContext.variant !== props.variant`.

The edge cases have to be handled by the user, I don't see another way.

**Comment by jguddas** [comment]
**Another idea:**

We could use `peer` and a `data-variant` attribute.

https://play.tailwindcss.com/Dl8H3CEOfv

**Comment by chongruei** [comment]
Hi @jguddas 

Thank you for your suggestion, which has provided me with a lot of inspiration. I have finally used an adjacent selector to check if the color and variant are the same. If they are the same, I will add a margin-left effect. 

This approach will cover all edge cases.

File: packages/components/button/stories/button-group.stories.tsx (hunk 3 above)
Comment (chongruei) [reply packages/components/button/stories/button-group.stories.tsx:102]: Thank you!

File: packages/components/button/stories/button-group.stories.tsx (hunk 2 above)
Comment (chongruei) [reply packages/components/button/stories/button-group.stories.tsx]: Thank you!

File: packages/components/button/src/use-button.ts (hunk 7 above)
Comment (chongruei) [reply packages/components/button/src/use-button.ts]: It's been removed the ```isIsolated``` condition.

File: packages/core/theme/src/components/button.ts (hunk 5 above)
Comment (chongruei) [reply packages/core/theme/src/components/button.ts]: Solved it.

File: packages/components/button/stories/button-group.stories.tsx (hunk 6 above)
Comment (chongruei) [reply packages/components/button/stories/button-group.stories.tsx]: Solved it.

File: packages/core/theme/src/components/button.ts [hunk 8]
```diff
@@ -340,8 +344,45 @@ const button = tv({
     // isInGroup / bordered / ghost
     {
       isInGroup: true,
-      variant: ["bordered", "ghost"],
-      class: "[&:not(:first-child)]:ml-[calc(theme(borderWidth.medium)*-1)]",
+      variant: ["ghost", "bordered"],
+      color: "default",
+      className:
+        "nextui-adjacent-default [&+.nextui-adjacent-default]:ml-[calc(theme(borderWidth.medium)*-1)]",
```
Comment (jguddas) [inline packages/core/theme/src/components/button.ts:350]: I would prefer using `data-color` and `data-variant` over `.nextui-adjacent-…`
```suggestion
        [
          "[[data-variant=bordered][data-color=default]_+_&]:ml-[calc(theme(borderWidth.2)*-1)]",
          "[[data-variant=ghost][data-color=default]_+_&]:ml-[calc(theme(borderWidth.2)*-1)]"
        ],
```

Comment (jguddas) [reply packages/core/theme/src/components/button.ts:350]: You could add a file with all these selectors in `../utils` to make it reusable.
```suggestion
        utils.collapseAdjacentVariantBorders.default,
```

Comment (jguddas) [reply packages/core/theme/src/components/button.ts:350]: When thinking about it using a class name also isn't that bad, it's pretty much just like using `peer`.

Comment (jguddas) [reply packages/core/theme/src/components/button.ts:350]: Something like this would also work.
```suggestion
        "[&+.border-medium.border-default]:ml-[calc(theme(borderWidth.medium)*-1)]",
```

Comment (chongruei) [reply packages/core/theme/src/components/button.ts]: Thank you @jguddas 

I have finally placed ```collapseAdjacentVariantBorders``` in the ```utils``` folder to make it reusable. This change also enables support for other border widths in the future.

**Review by jguddas — APPROVED** [review]

**Review by jrgarciadev — APPROVED** [review]
Huge thanks @chongruei @jguddas  ❤️

File: packages/core/theme/src/utils/classes.ts [hunk 9]
```diff
@@ -54,3 +54,16 @@ export const translateCenterClasses = [
...
+ * This object defines CSS classes for collapsing adjacent variant borders.
+ * It includes classes for different variants like default, primary, secondary, etc.
+ */
+export const collapseAdjacentVariantBorders = {
+  default: ["[&+.border-medium.border-default]:ml-[calc(theme(borderWidth.medium)*-1)]"],
+  primary: ["[&+.border-medium.border-primary]:ml-[calc(theme(borderWidth.medium)*-1)]"],
+  secondary: ["[&+.border-medium.border-secondary]:ml-[calc(theme(borderWidth.medium)*-1)]"],
+  success: ["[&+.border-medium.border-success]:ml-[calc(theme(borderWidth.medium)*-1)]"],
+  warning: ["[&+.border-medium.border-warning]:ml-[calc(theme(borderWidth.medium)*-1)]"],
+  danger: ["[&+.border-medium.border-danger]:ml-[calc(theme(borderWidth.medium)*-1)]"],
```
Comment (jguddas) [inline packages/core/theme/src/utils/classes.ts:68]:
```suggestion
  default: ["[&+.border-medium.border-default]:ms-[calc(theme(borderWidth.medium)*-1)]"],
  primary: ["[&+.border-medium.border-primary]:ms-[calc(theme(borderWidth.medium)*-1)]"],
  secondary: ["[&+.border-medium.border-secondary]:ms-[calc(theme(borderWidth.medium)*-1)]"],
  success: ["[&+.border-medium.border-success]:ms-[calc(theme(borderWidth.medium)*-1)]"],
  warning: ["[&+.border-medium.border-warning]:ms-[calc(theme(borderWidth.medium)*-1)]"],
  danger: ["[&+.border-medium.border-danger]:ms-[calc(theme(borderWidth.medium)*-1)]"],
```
